### PR TITLE
acme-client: 0.1.16 -> 0.2.4

### PR DESCRIPTION
**Pull request by ruuda**
###### Motivation for this change

A lot has happened since 0.1.16 was packaged. [The old upstream website](https://kristaps.bsd.lv/acme-client/) now says:

> Attention: this version of acme-client has been archived, as it now lives in OpenBSD base. If you're using the upstream version of this code, you're using old code! The live code, /usr/sbin/acme-client in OpenBSD, is well-maintained and current. See the CVS repository for current code.
>
> If you can't use OpenBSD, the current code-base can fairly easily be ported, but this is not an effort I have a need for. Please see the portable version's [GitHub](https://github.com/kristapsdz/acme-client-portable) README.md for hints. Thank you!

The linked repository is archived though, and its readme says:

> Attention acme-client-portable is no longer maintained. Since acme-client made its way into OpenBSD, I've only been using the base version.
>
> So if you're using this client, you're using old code! If you want to use the current version of acme-client, you'll need to use OpenBSD. Which I recommend doing anyway, of course.
>
> If you can't use OpenBSD, it's not a difficult challenge to port the OpenBSD acme-client to your operating system. However, be aware that it won't share the security mechanisms available to the downstream code. Regardless, if anybody would like to take a stab at this, I'd be happy to review your work!

Fortunately, [letsencrypt.org](https://letsencrypt.org/docs/client-options/#clients-c) links to https://github.com/graywolf/acme-client-portable, which does appear to be a maintained portable version of acme-client, so I took that repository as the new upstream.

It looks like that new acme-client-portable has been modified to build against OpenSSL, but that actually broke LibreSSL compatibility. I included a small patch to restore compatibility. I also submitted that patch upstream.

The new package produces a binary that runs, but I have not tested it extensively. Also, due to the changed upstream, and the portable version being a modified version of the OpenBSD version, it is hard to tell what has changed because of the version bump. I’ll try and see if I can diff the code from the tarballs.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @pmahoney

### CODE REVIEW

**Review by pmahoney — APPROVED** [review]
Nice work @ruuda . I was not aware of acme-client-portable. I had been considering proposing deleting this from nixpkgs..

**Comment by ruuda** [comment]
My compatibility patch has been merged upstream, and version 0.2.4 has been released that includes it, so I updated this PR to drop the patch and use 0.2.4 instead.

**Comment by ruuda** [comment]
Here is a diffstat of the changes from 0.1.16 to 0.2.4:

<details>
<summary>Diffstat</summary>

```
 /dev/null => acme-client-portable-0.2.4/.gitignore |    8 +
 .../null => acme-client-portable-0.2.4/CVS/Entries |   26 +
 .../CVS/Repository                                 |    1 +
 /dev/null => acme-client-portable-0.2.4/CVS/Root   |    1 +
 acme-client-portable-0.1.16/ChangeLog => /dev/null |  854 ---------------
 .../GNUmakefile => /dev/null                       |  190 ----
 .../null => acme-client-portable-0.2.4/Makefile.am |   58 ++
 /dev/null => acme-client-portable-0.2.4/NEWS       |   20 +
 /dev/null => acme-client-portable-0.2.4/README     |   88 ++
 .../acctproc.c                                     |  385 +++++--
 .../acme-client.1                                  |  463 ++------
 .../acme-client.conf.5                             |  177 ++++
 .../base64.c                                       |  147 ++-
 .../certproc.c                                     |  180 +---
 .../chngproc.c                                     |  185 ++--
 .../compat-setresgid.c => /dev/null                |   50 -
 .../compat-setresuid.c => /dev/null                |   50 -
 /dev/null => acme-client-portable-0.2.4/compat.c   |   31 +
 /dev/null => acme-client-portable-0.2.4/compat.h   |   26 +
 acme-client-portable-0.1.16/config.h => /dev/null  |   74 --
 .../configure.ac                                   |   54 +
 .../confs/example.conf                             |   12 +
 .../confs/test.conf                                |   12 +
 .../dbg.c                                          |    8 +-
 .../dnsproc.c                                      |   64 +-
 .../extern.h                                       |  112 +-
 .../fileproc.c                                     |  213 ++--
 .../http.c                                         |  690 +++++-------
 .../http.h                                         |   23 +-
 .../json.c                                         |  632 +++++++----
 .../rsa.c => acme-client-portable-0.2.4/key.c      |  103 +-
 .../rsa.h => acme-client-portable-0.2.4/key.h      |   11 +-
 .../keyproc.c                                      |  170 ++-
 .../m4/ax_check_bsd.m4                             |  126 +++
 .../m4/ax_check_openssl.m4                         |  124 +++
 .../main.c                                         |  694 ++++--------
 .../netproc.c                                      |  801 +++++++-------
 /dev/null => acme-client-portable-0.2.4/parse.h    |   87 ++
 /dev/null => acme-client-portable-0.2.4/parse.y    | 1100 ++++++++++++++++++++
 .../revokeproc.c                                   |  182 ++--
 .../sandbox-darwin.c => /dev/null                  |   64 --
 .../sandbox-null.c => /dev/null                    |   43 -
 .../sandbox-pledge.c => /dev/null                  |  110 --
 .../sandbox-seccomp.c => /dev/null                 |  278 -----
 .../scripts/prepare_builds                         |   33 +
 .../scripts/prepare_readme                         |    8 +
 .../scripts/update.sh                              |   29 +
 .../util-pledge.c => /dev/null                     |   65 --
 .../util-portable.c => /dev/null                   |   95 --
 .../util.c                                         |  127 +--
 50 files changed, 4363 insertions(+), 4721 deletions(-)
```

</details>

The diff itself is rather large, so I only skimmed through it superficially. A few notable things:

* 0.1.16 used Yoda conditions but 0.2.4 does not, and some formatting changes.
* 0.1.16 used what looks like a more elaborate BSD compatibility/sandbox layer, 0.2.4 uses `HAVE_PLEDGE` and `HAVE_UNVEIL` ifdef guards instead.
* Changes to the build setup.

**Comment by ruuda** [comment]
Is there anything I can do to help move this forward?

File: pkgs/tools/networking/acme-client/default.nix [hunk 1]
```diff
@@ -12,24 +15,27 @@ with stdenv.lib;
...
-    sha256 = "00q05b3b1dfnfp7sr1nbd212n0mqrycl3cr9lbs51m7ncaihbrz9";
+  src = fetchFromGitHub {
+    owner = "graywolf";
+    repo = "acme-client-portable";
+    rev = "v${version}";
+    sha256 = "1yq2lkrnjwjs0h9mijqysnjmr7kp4zcq1f4cxr9n1db7pw8446xb";
   };
 
-  buildInputs = [ libbsd libressl pkgconfig ]
+  buildInputs = [ autoconf automake bison libbsd libressl pkgconfig ]
```
Comment (c0bw3b) [inline pkgs/tools/networking/acme-client/default.nix]:
```suggestion
  nativeBuildInputs = [ autoreconfHook bison pkgconfig ];
  buildInputs = [ libbsd libressl ]
```

File: pkgs/tools/networking/acme-client/default.nix [hunk 2]
```diff
@@ -1,8 +1,11 @@
 { stdenv
+, fetchFromGitHub
+, autoconf
+, automake
```
Comment (c0bw3b) [inline pkgs/tools/networking/acme-client/default.nix]: replace both with just `autoreconfHook`

File: pkgs/tools/networking/acme-client/default.nix [hunk 3]
```diff
@@ -12,24 +15,27 @@ with stdenv.lib;
...
-  buildInputs = [ libbsd libressl pkgconfig ]
+  buildInputs = [ autoconf automake bison libbsd libressl pkgconfig ]
     ++ optional stdenv.isDarwin apple_sdk.sdk;
 
   CFLAGS = "-DDEFAULT_CA_FILE='\"${defaultCaFile}\"'";
 
   preConfigure = ''
+    autoreconf --install
     export PREFIX="$out"
   '';
```
Comment (c0bw3b) [inline pkgs/tools/networking/acme-client/default.nix]: The whole hook can become:
```nix
    makeFlags = [ "PREFIX=${placeholder "out"}" ];
```
and autoreconf is taken care of through nativeBuildInputs

File: pkgs/tools/networking/acme-client/default.nix [hunk 4]
```diff
@@ -12,24 +15,27 @@ with stdenv.lib;
...
+    repo = "acme-client-portable";
+    rev = "v${version}";
+    sha256 = "1yq2lkrnjwjs0h9mijqysnjmr7kp4zcq1f4cxr9n1db7pw8446xb";
   };
 
-  buildInputs = [ libbsd libressl pkgconfig ]
+  buildInputs = [ autoconf automake bison libbsd libressl pkgconfig ]
     ++ optional stdenv.isDarwin apple_sdk.sdk;
 
   CFLAGS = "-DDEFAULT_CA_FILE='\"${defaultCaFile}\"'";
```
Comment (c0bw3b) [inline pkgs/tools/networking/acme-client/default.nix]: We [hardcode the path to cacert CA bundle](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/libressl/default.nix#L40) in our LibreSSL package and it seems `DEFAULT_CA_FILE` is now ignored.
So I think this line (and `cacert` input) are not doing anything and could be removed.

**Review by c0bw3b — COMMENTED** [review]
Thanks for refreshing this package @ruuda 
I'm glad to learn there is a fork maintaining the portable variation for Linux systems

**Comment by ruuda** [comment]
Thanks for the review @c0bw3b, that was very educational. I addressed your comments, please take another look.

**Review by c0bw3b — APPROVED** [review]
Result of `nix-review pr 71853` [1](https://github.com/Mic92/nix-review)
<details>
  <summary>1 package were build:</summary>

  - acme-client
</details>

It builds and runs

**Pull request merged**